sslh-fork.c

```diff
@@ -178,6 +178,8 @@ void tcp_listener(struct listen_endpoint* endpoint, int num_endpoints, int activ
     while (1) {
         in_socket = accept(endpoint[active_endpoint].socketfd, 0, 0);
         CHECK_RES_RETURN(in_socket, "accept", /*void*/ );
+        if (in_socket < 0) continue;
```

Comment:
this will never be called, as CHECK_RES_RETURN() already checks for `in_socket == -1`.
Also, it shouldn't crash as the failed call to `accept()` is caught.
So, if you're seeing a crash, it would be some other reason (and I'd be interested in a stack analysis on the core dump).

OTOH, you're right that in the case of `accept()` failing, the TCP listener will stop, when really it might be a transcient error on that one connection. You can expand the CHECK_RES_RETURN to replace the `return` with a `continue` as you wrote.

Reply:
Hi yrutschle@, thank you for your reply! The crash we're experiencing is on 1.18, and the stack trace we see looks like this:

#6  0x00007fa73bfbd750 in __GI___chk_fail () at chk_fail.c:28
#7  0x00007fa73bfbed7a in __fdelt_chk (d=d@entry=-1) at fdelt_chk.c:25
#8  0x00005c8d00c7f70c in start_shoveler (in_socket=in_socket@entry=-1) at sslh-fork.c:80
#9  0x00005c8d00c7f9b6 in main_loop (listen_sockets=0x5c8d01ffee00, num_addr_listen=num_addr_listen@entry=1) at sslh-fork.c:162

While this is an old version, the processing pathway on top of tree and in 1.18 still seems similar. If I'm not mistaken, CHECK_RES_RETURN just prints on failure and returns void, but doesn't seem to prevent the loop from continuing to process this in_socket in the failed case. It seems that we would still jump into start_shoveler and attempt FD_SET on an invalid in_socket.

Perhaps your suggestion is to use the return from CHECK_RES_RETURN to decide to continue or not? If so, I can adjust my change to use the return from CHECK_RES_RETURN, let me know.

Reply:
Ah, I understand perfectly now. I was reading CHECK_RES_RETURN() as a function call that returned void, where in fact it's a macro that sticks the return into tcp_listener as you say, breaking out when accept() fails. This should handle the failure we see on 1.18, so let's close my request. Thanks very much for your explanation and help landing your fix!